### ISSUE TITLE
Improve matching on ADTs

### ISSUE THREAD

**Issue by denheck** [cauldron-lang/cauldron#24]
Make the following program work:
```
io := import("bifs:io")
print := io["print"]

adt {
  | Some(value)
  | None
}

maybe_one := Some(1)

match(maybe_one) {
  Some(value) -> print(value)
  None        -> print("No Value")
}
```

**Comment by denheck** [comment]
Still need to get empty ADTs like `None` to work in the `match`

**Comment by denheck** [comment]
Final fix for this is #29
